### ISSUE TITLE
reference fields in fhir-examples generated resources should include resource type

### ISSUE THREAD

**Issue by punktilious** [IBM/FHIR#1536]
**Describe the bug**
Many reference fields in the fhir-examples generated resources use short random string values like `IRm0J8cmzs`. Such values do not conform to any reference format defined in the FHIR specification.

**To Reproduce**
Build the examples package and inspect the generated resources:

- mvn clean install -f fhir-examples-generator
- mvn com.ibm.fhir:fhir-examples-generator:generate-examples -f fhir-examples
- view fhir-examples/src/main/resources/json/ibm/complete-mock/Patient-1.json

**Expected behavior**
Reference field values should be a URL or of the form `<resourceType>/<logicalId>`

**Comment by punktilious** [comment]
Implementation tip. In ModelSupport:

```
    /**
     * @param modelClass
     *            a model class which represents a FHIR resource or element
     * @param elementName
     *            the name of the reference element
     * @return a set of Strings which represent the the allowed target types for the reference
     */
    public static Set<String> getReferenceTargetTypes(Class<?> modelClass, String elementName) {
        ElementInfo elementInfo = getElementInfo(modelClass, elementName);
        if (elementInfo != null) {
            return elementInfo.getReferenceTypes();
        }
        return Collections.emptySet();
    }
```

**Comment by dermatologist** [comment]
Would love to help out and give it a go!

**Comment by prb112** [comment]
> Would love to help out and give it a go!

Fantastic.  We're happy to coach, and we are generally available vai Zulip on #ibm  https://chat.fhir.org/#narrow/stream/212434-ibm

**Comment by dermatologist** [comment]
Thank you. I think to reproduce it should be in the file fhir-examples-generator folder. right?

* mvn clean install -f fhir-examples-generator
* mvn com.ibm.fhir:fhir-examples-generator:generate-examples -f **fhir-examples-generator**/pom.xml
* view fhir-examples/src/main/resources/json/ibm/complete-mock/Patient-1.json

**Comment by lmsurpre** [comment]
you may need to build the whole project first since the fhir-examples-generator is currently configured to depend on the 4.4.1-SNAPSHOT version of our other modules.
* mvn clean install -f fhir-examples
* mvn clean install -f fhir-parent -DskipTests

**Comment by prb112** [comment]
Unfortunately, no - the examples generator builds JSON files which are destined for `fhir-examples` not `fhir-examples-generator`

And yes, @lmsurpre is correct (dependencies must be built first)

**Comment by dermatologist** [comment]
Thanks, @prb112 and @lmsurpre 

Just adding a few things I found:

I see this in the ExamplesPlugin (and hence my previous comment):

```
        // Only runs for the fhir-examples-generator project since that is where the definitions directory exists, short-circuits otherwise.
        if (mavenProject.getArtifactId().contains("fhir-examples-generator")) {
```
It works for me when I use *-f fhir-examples-generator* (only), and most references to FHIR resources are in the correct format (Patient/XXXX, Encounter/XXXX, Organization/XXXX etc.) in the generated files, but some like Patient.generalPractitioner are not.

This may be related to how they are defined in the definitions/profiles-resources.json. I noticed that all resources not properly referenced have a '-' in their definitions:
Patient.general-practitioner
Claim.care-team

**Comment by prb112** [comment]
Thanks @dermatologist 

Sure - it may be setup like that, and then there would be an extra step to ensure the contents are end up in the fhir-examples project. 

Your observations sounds correct with regards to the generated content.  I think Lee is best to answer the detail of the question.

Paul

**Comment by prb112** [comment]
Moved to Sprint 20

**Comment by lmsurpre** [comment]
> I noticed that all resources not properly referenced have a '-' in their definitions

I don't think thats it.  For example, see `Goal.addresses` and `Goal.outcomeReference`.

With respect to
> Only runs for the fhir-examples-generator project

I think that was a gap in my own understanding of how this should work.
Typically I just run the ExamplesGenerator main and then copy paste the updated examples over.
However, the end goal was to make this work similar to our `fhir-tools` plugin that generates the model classes directly to `fhir-model`.
To this end, I've just opened a pull request that should clean this up significantly: https://github.com/IBM/FHIR/pull/1656
If you're able to rebase onto that, you can invoke the examples generator via `build/gen-examples.sh` (or do something similar manually on windows) and it should properly update the examples in the right spot of the `fhir-examples` project.

**Comment by lmsurpre** [comment]
@dermatologist this one turned out to be more involved than we had first thought.  
After a deeper investigation, I found two different cases where we weren't setting the reference type properly:
1. when the reference was a choice element
2. when the reference was a repeating

I went ahead and pushed some changes to a pull request at https://github.com/IBM/FHIR/pull/1667/commits which should address the root of the issue.
If there's anything we could do to make it easier for potential contributors like yourself, we'd love your feedback on that.  Thanks!
